Skip user activation check in requestStorageAccess if frame ever gets NoModificationAllowedError

szewai opened the pull request:
#### 2173fe1250ad9034b35bac7d1495e7ae2fdcacdc
<pre>
Skip user activation check in requestStorageAccess if frame ever gets NoModificationAllowedError
<a href="https://bugs.webkit.org/show_bug.cgi?id=269557">https://bugs.webkit.org/show_bug.cgi?id=269557</a>

Reviewed by Alex Christensen.

If an iframe gets NoModificationAllowedError on requestStorageAccess, it means the request is granted but iframe needs
to reload to get the access. However, after reloading, the iframe would lose user activation, which stops it from
invoking requestStorageAccess again (Storage Access API requires user activation). To solve this, frame now stores user
activation for domains that ever get NoModificationAllowedError. If document does not have active user activation, but
its domain has user activation based on frame&apos;s records, then we will not reject the storage access request.

Updated test: http/tests/storageAccess/request-throw-exception-on-grant-until-reload.html. The test is modified to
invoke requestStorageAccess automatically (without another click) after reload.

* LayoutTests/http/tests/storageAccess/request-throw-exception-on-grant-until-reload.html:
* LayoutTests/http/tests/storageAccess/resources/request-throw-exception-on-grant-until-reload-iframe.html:
* Source/WebCore/dom/DocumentStorageAccess.cpp:
(WebCore::DocumentStorageAccess::requestStorageAccessQuickCheck):
(WebCore::DocumentStorageAccess::requestStorageAccess):
* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::storageAccessExceptionReceivedForDomain):
(WebCore::LocalFrame::requestSkipUserActivationCheckForStorageAccess):
* Source/WebCore/page/LocalFrame.h:

Canonical link: <a href="https://commits.webkit.org/274887@main">https://commits.webkit.org/274887@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fc5bf4bf50e764344cbff324d668ee408d33c327

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40254 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19266 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42632 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42799 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/36343 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/42561 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22193 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16595 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/33438 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40828 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16225 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/34743 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/14011 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/14055 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35696 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44077 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36502 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/36026 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/39773 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/15069 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12350 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/38052 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16688 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/34979 "Passed tests") | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/9046 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/16737 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/16332 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -4385,6 +4385,9 @@ void FrameLoader::dispatchDidCommitLoad(std::optional<HasInsecureContent> initia
         return;
 
+    if (m_frame.ptr())
+        auto url = m_frame->document() ? m_frame->document()->url() : URL { };
```
This is unused.

achristensen07 commented on Source/WebCore/page/LocalFrame.cpp:
```diff
@@ -1290,6 +1290,24 @@ CheckedRef<FrameSelection> LocalFrame::checkedSelection() const
+        return false;
+
+    // We currently only allow the domain to skip check once.
+    bool isActivated = LocalDOMWindow::hasTransientActivation(iter->value);
```
I think it is incorrect to store the time and call LocalDOMWindow::hasTransientActivation with that time.  It doesn't matter how long it takes for the frame to reload.  It should skip the check once after reload even if it takes a full minute, which won't be uncommon in some network conditions.

achristensen07 commented on Source/WebCore/page/LocalFrame.cpp:
```diff
@@ -1290,6 +1290,22 @@ CheckedRef<FrameSelection> LocalFrame::checkedSelection() const
+    if (iter == m_storageAccessExceptionDomains.end())
+        return false;
+
+    // We currently only allow the domain to skip check once.
```
Maybe remove "currently"

johnwilander reviewed (commented):
Judging from the title, this sounds like a leaky abstraction. We should not tie the relaxed user gesture requirement to an error type just because we use that error type in a certain way today. Someone may come along and trigger that error for another reason and not understand that they now relaxed the user gesture requirement for the Storage Access API.

The relaxation needs to depend on something that explicitly has to do with the Storage Access API, not the error type emitted to the web developer.

szewai committed:
Skip user activation check in requestStorageAccess if frame ever gets NoModificationAllowedError
https://bugs.webkit.org/show_bug.cgi?id=269557

Reviewed by Alex Christensen.

If an iframe gets NoModificationAllowedError on requestStorageAccess, it means the request is granted but iframe needs
to reload to get the access. However, after reloading, the iframe would lose user activation, which stops it from
invoking requestStorageAccess again (Storage Access API requires user activation). To solve this, frame now stores user
activation for domains that ever get NoModificationAllowedError. If document does not have active user activation, but
its domain has user activation based on frame's records, then we will not reject the storage access request.

Updated test: http/tests/storageAccess/request-throw-exception-on-grant-until-reload.html. The test is modified to
invoke requestStorageAccess automatically (without another click) after reload.

* LayoutTests/http/tests/storageAccess/request-throw-exception-on-grant-until-reload.html:
* LayoutTests/http/tests/storageAccess/resources/request-throw-exception-on-grant-until-reload-iframe.html:
* Source/WebCore/dom/DocumentStorageAccess.cpp:
(WebCore::DocumentStorageAccess::requestStorageAccessQuickCheck):
(WebCore::DocumentStorageAccess::requestStorageAccess):
* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::storageAccessExceptionReceivedForDomain):
(WebCore::LocalFrame::requestSkipUserActivationCheckForStorageAccess):
* Source/WebCore/page/LocalFrame.h:

Canonical link: https://commits.webkit.org/274887@main
